src/core/lib/iomgr/ev_apple.cc

```diff
@@ -0,0 +1,354 @@
+
+// pollset implementation
+
+static void pollset_global_init(void) {
+  gGlobalRunLoopContext = new GlobalRunLoopContext;
+
+  grpc_apple_register_read_stream_impl =
+      grpc_apple_register_read_stream_run_loop;
+  grpc_apple_register_write_stream_impl =
+      grpc_apple_register_write_stream_run_loop;
+
+  grpc_core::MutexLock lock(&gGlobalRunLoopContext->mu);
+  gGlobalRunLoopThread =
+      new grpc_core::Thread("apple_ev", GlobalRunLoopFunc, nullptr);
+  gGlobalRunLoopThread->Start();
+  while (gGlobalRunLoopContext->run_loop == NULL)
+    gGlobalRunLoopContext->init_cv.Wait(&gGlobalRunLoopContext->mu);
+}
+
+static void pollset_global_shutdown(void) {
+  {
+    grpc_core::MutexLock lock(&gGlobalRunLoopContext->mu);
+    gGlobalRunLoopContext->is_shutdown = true;
+    CFRunLoopStop(gGlobalRunLoopContext->run_loop);
+  }
+  gGlobalRunLoopThread->Join();
+  delete gGlobalRunLoopThread;
+  delete gGlobalRunLoopContext;
+}
+
+/// The caller must acquire the lock GrpcApplePollset.mu before calling this
+/// function. The lock may be temporarily released when waiting on the condition
+/// variable but will be re-acquired before the function returns.
+///
+/// The Apple pollset simply waits on a condition variable until it is kicked.
+/// The network events are handled in the global run loop thread. Processing of
+/// these events will eventually trigger the kick.
+static grpc_error* pollset_work(grpc_pollset* pollset,
+                                grpc_pollset_worker** worker,
+                                grpc_millis deadline) {
+  GRPC_POLLING_TRACE("pollset work: %p, worker: %p, deadline: %" PRIu64,
+                     pollset, worker, deadline);
+  GrpcApplePollset* apple_pollset =
+      reinterpret_cast<GrpcApplePollset*>(pollset);
+  GrpcAppleWorker actual_worker;
+  if (worker) {
+    *worker = reinterpret_cast<grpc_pollset_worker*>(&actual_worker);
+  }
+
+  if (apple_pollset->outstanding_kick) {
+    // Process the outstanding kick and reset the flag. Do not block.
+    apple_pollset->outstanding_kick = false;
+  } else {
+    // Block until kicked, timed out, or the pollset shuts down.
+    apple_pollset->workers.push_front(&actual_worker);
+    auto it = apple_pollset->workers.begin();
+
+    while (!actual_worker.kicked && !apple_pollset->is_shutdown) {
+      if (actual_worker.cv.Wait(
+              &apple_pollset->mu,
+              grpc_millis_to_timespec(deadline, GPR_CLOCK_REALTIME))) {
+        // timed out
+        break;
+      }
+    }
+
+    apple_pollset->workers.erase(it);
+
+    // If the pollset is shut down asynchronously and this is the last pending
+    // worker, the shutdown process is complete at this moment and the shutdown
+    // callback will be called.
+    if (apple_pollset->is_shutdown && apple_pollset->workers.empty()) {
+      grpc_core::ExecCtx::Run(DEBUG_LOCATION, apple_pollset->shutdown_closure,
+                              GRPC_ERROR_NONE);
+    }
+  }
+
+  return GRPC_ERROR_NONE;
+}
+
```

Comment:
Please add a comment to explain which mutex must be held before calling kick_worker().